video: dcmi: add endpoint usage in dcmi & update dma property in dtsi

avolmat-st opened the pull request:
This PR update the DCMI driver / bindings and all users of this driver in order to rely on the video-interfaces bindings rather that having its own properties.
In a second step, considering that the DMA settings are closely linked to the IP itself and the way it is connected internally, I propose to move the dma property from within the board dts into the dtsi so that each board / shield overlay do not have to copy again the same property. As a matter of fact the dcmi driver is also setting most of the DMA settings from within the driver itself.

Commit:
dts: bindings: stm32-dcmi: use endpoint based properties

Update the bindings of the stm32-dcmi driver rely on
properties described within the endpoints and already
detailed within the video-interfaces.yaml.

With that, several properties located at the node root
are now moved into the port / endpoint:

  sensor -> endpoint: remote-endpoint-label
  vsync-active -> endpoint: vsync-active
  hsync-active -> endpoint: hsync-active
  pixelclk-active -> endpoint: pclk-sample
  bus-width -> endpoint: bus-width

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: stm32: dcmi: perform config based on endpoint properties

Perform sensor interface properties parsing based on values
retrieved via the endpoint rather than the root of the node.
Use DT_PROP_OR to ensure proper configuration of optional
settings.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: weact_ov2640_cam_module: use endpoint based properties

Update overlay following usage of video-interfaces based
endpoint properties by the dcmi driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: st_b_cams_omv_mb1683: use endpoint based properties

Update overlay following usage of video-interfaces based
endpoint properties by the dcmi driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: arduino_nicla_vision: use endpoint based properties

Update overlay following usage of video-interfaces based
endpoint properties by the dcmi driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: st: h7: move dma property of dcmi in stm32h7.dtsi

Usage of dma is mandatory for the dcmi and this property is
tightly coupled with the soc itself since the configuration of
the dma depends on the source/destination, and the request line
is also fixed for an ip.
Instead of having to always have the dma property part of the
board or shield dts/overlay, add the dma property into the
dcmi node of the stm32h7.dtsi.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: bindings: video: dcmi: remove the dma in board dts example

With the addition of the dma property within the soc dtsi, it is
no more necessary to add it within the board dts.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
Sadly I do not have the HW to test it right now so I have only been able to compile test it. Could someone having those setup perform a sanity check that I didn't broke anything ?

ngphibang commented:
cc @CharlesDias so that he is aware of this to avoid duplicate work. I think part of this has been started in https://github.com/zephyrproject-rtos/zephyr/pull/85452 but this one is more complete.

Could you add an overlay and test case in video's build_all tests as well so that the DCMI driver get in the CI test (https://github.com/zephyrproject-rtos/zephyr/issues/85178) ?

Thanks !

ngphibang commented on dts/bindings/video/st,stm32-dcmi.yaml:
```diff
@@ -11,23 +11,21 @@ description: |
-      hsync-active = <0>;
-      vsync-active = <0>;
-      pixelclk-active = <1>;
-      capture-rate = <1>;
```
Would you support of changing framerate with the video API after removing this propperty ?

avolmat-st replied:
Actually I only removed it from the example, it is still present in the bindings itself. However it is optional and within the driver I took care of having DT_PROP_OR with a default to 1, aka no frame skip. So since this is the default (and actually quite natural), I thought it is simply not necessary to mention it here.

Regarding the already started work, sorry, I wasn't aware about it hence I did that. Sure I can add the remaining bits within the build_all tests.

avolmat-st replied:
Ok I just had a look at the PR #85452 now so I understand better your point. The DCMI IP is actually capable of frame dropping. The capture-rate mentioned here is actually more like a drop rate. 1 means keep 1/1 frames (aka all), 2 means keep 1 frame out of 2 and 4 would mean keep 1 frame out of 4.
So, basically, it would be possible to map this on the framerate API by first checking the framerate up the pipeline to figure out (if possible) the source framerate (might have to go quite far there might be a bridge or so in the middle), and then compute the rate between the framerate expected by the caller compared to the framerate of the source. Not an impossible thing. I can give it a shot, however again, I will only be able to blindly check it since I don't have currently a working setup with the DCMI on it.

CharlesDias replied:
Hi, @avolmat-st. Thanks for your contribution.

@ngphibang, it's ok for me to continue with this PR instead of https://github.com/zephyrproject-rtos/zephyr/pull/85452. :-)

I tested this PR with the `west build -p -b stm32h7b3i_dk --shield st_b_cams_omv_mb1683 samples/drivers/video/capture_to_lvgl/ `and worked!

Commit:
video: stm32-dcmi: correct get/set fmt handling

This commit mainly correct the get/set format handling and how
DCMI format is stored within the driver.  struct video_format
within the data structure is used to store the format.
Reworked way to handle get format to avoid calling the sensor
set_fmt whenever performing the get_fmt.
Slightly adjusted code to as much as possible reuse return
values provided by functions.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: stm32-dcmi: implement frame interval handling

Implement the video API frame interval handling in order
to control the framerate of capture.

This allow to remove the capture-rate DT property as well.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
tests: drivers: build_all: video: add testcase of stm32 dcmi

Add a testcase for building the stm32 dcmi driver on all
currently supported platforms / shields.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang commented:
> Last commit adds a testcase within test/drivers/build_all/video in order to test build the dcmi. For that purpose I am using one of the board and its shield so that dcmi get enabled and thus build. I ran this with twister and it seems to build, however couldn't figure out how to see the full log and really ensure that dcmi get enabled in this configuration.
> 
> @ngphibang is that what you were expecting regarding the testcase ? I saw that in case of the NXP board there is an overlay within the build_all folder, it seems to me that if really shield get enabled then it should be enough to enable the build of the driver no ?

Yes, the driver get built by just simply adding the shield as an extra args in the test case if we have only one platform_allow.

The NXP's case is a bit different. We have CSI which is common on 2 platforms, and if I remember well, when I do:

```
  drivers.video.mcux_csi.build:
    platform_allow:
      - mimxrt1064_evk
      - mimxrt1170_evk/mimxrt1176/cm7
    extra_args:
      - platform:mimxrt1064_evk:SHIELD=dvp_fpc24_mt9m114
      - platform:mimxrt1170_evk/mimxrt1176/cm7:SHIELD=nxp_btb44_ov5640
```

It didn't get built because "platform" is not taken into account. I didn't know why.

Another reason is that I see this in the app.overlay of the build_all video:

```
/*
 * Copyright (c) 2022, Kumar Gala <galak@kernel.org>
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Application overlay for testing driver builds
 *
 * Names in this file should be chosen in a way that won't conflict
 * with real-world devicetree nodes, to allow these tests to run on
 * (and be extended to test) real hardware.
 */
```

So, I supposed that for CI testing, we should declare "fake" nodes to avoid conflicting with real-world device tree node.
So, to simplify things for two above reasons, I made the overlay.

I am not expert in test, so if you find some thing, I am willing to learn.

ngphibang reviewed (dismissed):
Looks very nice to me ! Thank you for doing the migration of the driver.

Waiting for your finding on the build_all video test (to see if overlay is the right thing to do or the simple extra args).

josuah commented on tests/drivers/build_all/video/testcase.yaml:
```diff
@@ -22,3 +22,6 @@ tests:
   drivers.video.mcux_smartdma.build:
     platform_allow:
       - frdm_mcxn947/mcxn947/cpu0
+  drivers.video.stm32_dcmi.build:
+    platform_allow: stm32h7b3i_dk
+    extra_args: SHIELD=st_b_cams_omv_mb1683
```
There it is I think:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/14917714935/job/41907115938#step:12:348

josuah commented:
It still works!

Using the [patch soup](https://github.com/zephyrproject-rtos/zephyr/discussions/89691):
```
git remote add tvai https://github.com/tinyvision-ai-inc/zephyr
git fetch tvai
git cherry-pick tvai/pr-usb-uvc~9..tvai/pr-usb-uvc
west build -b arduino_nicla_vision/stm32h747xx/m7 samples/subsys/usb/uvc/
west flash
```

avolmat-st commented:
> Looks very nice to me ! Thank you for doing the migration of the driver.
> 
> Waiting for your finding on the build_all video test (to see if overlay is the right thing to do or the simple extra args).

I did further tests using twister and managed to have properly working platform_allow and extra_args including SHIELD statement. Since the DCMI is already enabled on several platforms, I've added it. Normally this would make 3 plateforms however mini_stm32h743 is failing to build using the shield since the shield is referring to application specific CONFIG. So it fact the conf of the shield has to be reworked and the conf should be done within the application side. To avoid blocking this I've thus not added the mini_stm32h743 into the test. It should be added later on once fixed.

Important part was to actually mention the full name for the platform. Example:

     platform_allow:
       - stm32h7b3i_dk/stm32h7b3xx
       - arduino_nicla_vision/stm32h747xx/m7
      - mini_stm32h743
     extra_args:
       - platform:stm32h7b3i_dk/stm32h7b3xx:SHIELD=st_b_cams_omv_mb1683
      - platform:mini_stm32h743/stm32h743xx:SHIELD=weact_ov2640_cam_module

hence not only stm32h7b3i_dk, but stm32h7b3i_dk/stm32h7b3xx. With that done, twister is properly building for the 3 platforms, including the shield in case of stm32h7b3i and mini_stm32h743 and no shield in case of arduino_nicla_vision.

With that in place, my feeling is that there is no need to add specific overlay within the test since this would mean duplication of code hence maintenance issues. Of course if there is no shield available for a platform we will have to add overlay in the test section.

josuah reviewed (dismissed):
The modifications LGTM.

This issue was opened regarding the conflicting CONFIG_VIDEO_HFLIP=y CONFIG_VIDEO_VFLIP=y

- https://github.com/zephyrproject-rtos/zephyr/issues/89783

ngphibang commented:
> With that in place, my feeling is that there is no need to add specific overlay within the test since this would mean duplication of code hence maintenance issues. Of course if there is no shield available for a platform we will have to add overlay in the test section.

I just tested again with the extra_args. It always NOT get the CSI driver built (I made some thing wrong in the driver code but CI still passed and it says platform is not taken into account).

```
west build -p -b mimxrt1170_evk/mimxrt1176/cm7 tests/drivers/build_all/video -T drivers.video.mcux_csi.build

CMake Warning:
  Manually-specified variables were not used by the project:

    platform

  drivers.video.mcux_csi.build:
    platform_allow:
      - mimxrt1064_evk
      - mimxrt1170_evk/mimxrt1176/cm7
    extra_args:
      - platform:mimxrt1064_evk:SHIELD=dvp_fpc24_mt9m114
      - platform:mimxrt1170_evk/mimxrt1176/cm7:SHIELD=nxp_btb44_ov5640
```

I am not sure why it works in your case. Maybe because the DCMI driver is always enabled by default in ST boards and NOT by the shields ? If so, it is normal.

avolmat-st commented:
> > With that in place, my feeling is that there is no need to add specific overlay within the test since this would mean duplication of code hence maintenance issues. Of course if there is no shield available for a platform we will have to add overlay in the test section.
> 
> I just tested again with the extra_args. It always NOT get the CSI driver built (I made some thing wrong in the driver code but CI still passed and it says platform is not taken into account).
> 
> ```
> west build -p -b mimxrt1170_evk/mimxrt1176/cm7 tests/drivers/build_all/video -T drivers.video.mcux_csi.build
> 
> CMake Warning:
>   Manually-specified variables were not used by the project:
> 
>     platform
> 
>   drivers.video.mcux_csi.build:
>     platform_allow:
>       - mimxrt1064_evk
>       - mimxrt1170_evk/mimxrt1176/cm7
>     extra_args:
>       - platform:mimxrt1064_evk:SHIELD=dvp_fpc24_mt9m114
>       - platform:mimxrt1170_evk/mimxrt1176/cm7:SHIELD=nxp_btb44_ov5640
> ```
> 
> I am not sure why it works in your case. Maybe because the DCMI driver is always enabled by default in ST boards and NOT by the shields ? If so, it is normal.

the platform name aren't correct I think. I just tested with the patch below applied and with that done it is well applying the shields during build. In case of the mimxrt1170 it even complian that there are 2 same label for the sensor since now it is via the shield and via the test overlay:

diff --git a/tests/drivers/build_all/video/testcase.yaml b/tests/drivers/build_all/video/testcase.yaml
index 57e74e80168..bca36e73876 100644
--- a/tests/drivers/build_all/video/testcase.yaml
+++ b/tests/drivers/build_all/video/testcase.yaml
@@ -15,6 +15,9 @@ tests:
     platform_allow:
       - mimxrt1064_evk
       - mimxrt1170_evk/mimxrt1176/cm7
+    extra_args:
+      - platform:mimxrt1064_evk/mimxrt1064:SHIELD=dvp_fpc24_mt9m114
+      - platform:mimxrt1170_evk@B/mimxrt1176/cm7:SHIELD=nxp_btb44_ov5640
   drivers.video.mcux_mipi_csi2rx.build:
     platform_allow:
       - mimxrt1170_evk/mimxrt1176/cm7

ngphibang commented:
> the platform name aren't correct I think. I just tested with the patch below applied and with that done it is well applying the shields during build. In case of the mimxrt1170 it even complian that there are 2 same label for the sensor since now it is via the shield and via the test overlay:

In fact, `mimxrt1170_evk` is now default to `mimxrt1170_evk@B` so `mimxrt1170_evk/mimxrt1176/cm7` and `mimxrt1170_evk@B/mimxrt1176/cm7` can be used alternatively.

And I did exactly the same as you did but still always got the same result. Tested  more throroughly:

`west twister -i -T tests/drivers/build_all/video` does detect 2 errors:

```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests/drivers/build_all/video/drivers.video.mcux_csi.build on mimxrt1064_evk/mimxrt1064 error (Build failure - error: videodevice.h: No such file or directory)
INFO    - 2) tests/drivers/build_all/video/drivers.video.mcux_mipi_csi2rx.build on mimxrt1170_evk@B/mimxrt1176/cm7 error (Build failure - error: videodevice.h: No such file or directory)
INFO    - 
INFO    - To rerun the tests, call twister using the following commandline:
INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
INFO    - 
INFO    - west twister -p mimxrt1170_evk@B/mimxrt1176/cm7 -s tests/drivers/build_all/video/drivers.video.mcux_mipi_csi2rx.build
INFO    - or with west:
INFO    - west build -p -b mimxrt1170_evk@B/mimxrt1176/cm7 tests/drivers/build_all/video -T drivers.video.mcux_mipi_csi2rx.build
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - Run completed
```

But note that the 2nd error is detected on the mipi_csi2rx test, not the csi test. I checked more :

- `west twister -p mimxrt1064_evk/mimxrt1064 -s tests/drivers/build_all/video/drivers.video.mcux_csi.build`
=> it detected the error.

But:

- `west build -p -b mimxrt1064_evk/mimxrt1064 tests/drivers/build_all/video -T drivers.video.mcux_csi.build`
=> it canNOT detect the error while these two commands can be used alternatively as said in the above logs ???. It is very weid to me.

And both:

- `west twister -p mimxrt1170_evk@B/mimxrt1176/cm7 -s tests/drivers/build_all/video/drivers.video.mcux_csi.build`
- `west build -p -b mimxrt1170_evk@B/mimxrt1176/cm7 tests/drivers/build_all/video -T drivers.video.mcux_csi.build`

canNOT detect the error.

ngphibang reviewed (dismissed):
Approved as the ongoing discussion is not specific to ST boards.

erwango reviewed (changes requested):
Given the binding change and impact on out of tree users, please update `doc/releases/migration-guide-4.2.rst`

Commit:
doc: releases: add note regarding dcmi binding update

Add note regarding the move of the video/dcmi driver
to the usage of endpoint based video-interfaces
bindings.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -487,6 +487,12 @@ xSPI
+Video
+=====
+
+* On STM32 devices, the DCMI driver now relies on endpoint based video-interfaces.yaml bindings
+  for sensor interface properties (such as bus width and synchronization signals). The capture-rate
+  property has been replaced by the usage of the frame interval (frmival) API.
```
Let's add clickable references to help users migrating more easily
```suggestion
* On STM32 devices, the DCMI driver (:dtcompatible:`st,stm32-dcmi`) now relies on endpoint based
  video-interfaces.yaml bindings for sensor interface properties (such as bus width and synchronization signals).
  Also the ``capture-rate`` property has been replaced by the usage of the frame interval API :c:func:`video_set_frmival`. 
  See (:github:`89627`).
```

erwango reviewed (changes requested):
Otherwise LGTM

avolmat-st commented:
> Otherwise LGTM

migration-guide message formatting update.

avolmat-st commented on dts/bindings/video/st,stm32-dcmi.yaml:
```diff
@@ -11,23 +11,21 @@ description: |
       port {
         dcmi_ep_in: endpoint {
-          remote-endpoint = <&ov2640_ep_out>;
+          remote-endpoint-label = "ov2640_ep_out";
```
Yes this is a limitation currently but not related to the DCMI driver itself:

from: dts/bindings/video/video-interfaces.yaml

  Two 'endpoint' nodes must be linked with each other via their 'remote-endpoint'
  phandles. However, Zephyr does not allow circular dependency, so direct phandle
  references are currently not possible. A 'remote-endpoint-label' string is used
  instead to be able to specify, at least, the label of the peer remote-endpoint.
  For example:

          source: endpoint {
              compatible = "zephyr,video-interfaces";
              remote-endpoint-label = "sink";
          };

          sink: endpoint{
              compatible = "zephyr,video-interfaces";
              remote-endpoint-label = "source";
          };

josuah replied:
The issue about it is here:
- https://github.com/zephyrproject-rtos/zephyr/issues/57708
